### ISSUE TITLE
Bug in ini library

### ISSUE THREAD

**Issue by mcblum** [DannyBen/bashly#556]
### Bashly Version

Latest Ruby Gem

### Description

What I'm seeing is that sometimes values are added or deleted from the wrong section. I was banging my head against the wall trying to figure out what was going on and then I ran a test operation and asked it to delete `tenant.DATA_PATH` from one section but it, in fact, deleted it from another. This has caused some serious issues.

Also, sometimes the ini file is just blank altogether, which I really don't understand. 

### Reproduction Steps

Basically I'm looping through a bunch of tenant folders, reading their INI files, and sometimes updating values. The updates are frequently getting messed up.

### Actual Behavior

Keys are randomly deleted from the ini file, sometimes the ini file is written blank.

### Expected Behavior

That loading the INI and changing a value would work as expected

**Comment by DannyBen** [comment]
Without a minimal, simple way to reproduce the problem, there is not much I can do.

**Comment by mcblum** [comment]
I'm working on it right now! I would guess that means this isn't something that people are reporting often?

**Comment by DannyBen** [comment]
No reports. You can also take a look at the code for the INI library, it really doesn't do much.

https://github.com/DannyBen/bashly/blob/c0fd6f25b25065375638e03e5d61804cef4563ff/lib/bashly/libraries/ini/ini.sh#L56-L88

**Comment by Nsomnia** [comment]
Do a vs step by step debug, bang in some sleep/echo/printf commands in bashly, bashdb/similar, or whatever you might be using as libs/src with related ideas, to figure out when *and* why it happens?

**Comment by DannyBen** [comment]
I am not following.

If it is a bug in the INI library in bashly, it should be easily reproducible.
If the code to reproduce is it complicated, you need to uncomplicate it, and be sure it is a bug in the INI library provided with bashly, and not elsewhere in your script.

If it is inconsistent, then I suspect a race condition.

Keep in mind that the code libraries provided with bashly are provided as a basis for you to build on. It is not the primary purpose of bashly to provide you with bash snippets. That said, if there is an obvious bug with any of the provided libraries, I would like to have them fixed.

**Comment by mcblum** [comment]
>If it is inconsistent, then I suspect a race condition.

This is exactly what I was thinking, but I can't find where that would be happening. I'm finding it hard to reproduce because IRL there are a lot of variables, so it's a lot of stuff to fake. The general process for us is

- loop through tenants
- for each tenant, load their ini
- read/write some config
- continue on

I suspect that's where the issue might be. I also hear what you're saying about it not being a Bashly thing, and that makes sense.

**Comment by DannyBen** [comment]
As you can see from the first line in the `ini_save()` function:

https://github.com/DannyBen/bashly/blob/c0fd6f25b25065375638e03e5d61804cef4563ff/lib/bashly/libraries/ini/ini.sh#L56C10-L57C8

it relies on a global variable - to set/get values.

This means that if it is called more than once at the same time, you will see issues. It was not designed for it.
See if there is any chance your code calls it with overlapping calls.

**Comment by DannyBen** [comment]
See if this helps:

Testing multiple INI files loaded in the same script causes an issue now. Here is the reproduction:

```bash
ini_load "test.ini"

for key in $(ini_keys); do
  echo "- $key = ${ini[$key]}";
done

unset ini

ini_load "test2.ini"

for key in $(ini_keys); do
  echo "- $key = ${ini[$key]}";
done
```

This will "blend" all configuration from all loaded INI files, unless you call "unset ini" as I did in the script.
Could this be the same issue you see?

**Comment by DannyBen** [comment]
In general, you should consider this:

The few bashly libraries you can add with `bashly add` are added to *your* folder by design.
The reason being, that you should treat them almost as if they are your own code. 

- If they are not to your liking - change them.
- If you see a bug in them - fix it inside their own functions
- If the changes/fixes are something that might be interesting to others, we can discuss updating the base libraries shipped with bashly.

In the case you discovered - it seems like adding `unset ini`, or providing means for specifying the name of the variable (or both) - might be of interest to others.

So - I am waiting to hear if adding `unset ini` inside the `ini_load` function seems to be solving your issue. If so, we can proceed.

**Comment by DannyBen** [comment]
Implemented in version 1.2.3
